custom/icds_reports/utils/__init__.py

```diff
@@ -1150,23 +1163,90 @@ def create_thr_report_excel_file(excel_data, data_type, month, aggregation_level
+    main_headers = ['State', 'District', 'Block', 'Sector', 'Awc Name', 'AWW Name', 'AWW Phone No.',
+                   'Total No. of Beneficiaries eligible for THR',
+                   'Total No. of beneficiaries received THR in given month',
+                   'Total No of Pictures taken by AWW']
+    headers.extend(main_headers[aggregation_level:])
+    beneficiary_type_columns = []
+    secondary_headers = []
+    if report_type != 'consolidated':
+        beneficiary_type_columns = [
+            'Pregnant women',
+            'Lactating women',
+            'Children (0-3 years)'
+        ]
+    if report_type == 'days_beneficiary_wise':
+        secondary_headers = ['Not provided',
+                             'Provided for 1-7 days',
+                             'Provided for 8-14 days',
+                             'Provided for 15-20 days',
+                             'Provided for 21-24 days',
+                             'Provided for at least 25 days (>=25 days)']
+
+    def set_beneficiary_columns(start_column_index, end_column_index, row):
+        for i in range(end_column_index - start_column_index + 1):
+            cell = "{}{}".format(columns[start_column_index + i], row)
+            worksheet[cell].fill = purple_fill
+            worksheet[cell].border = thin_border
+            worksheet[cell].font = bold_font
+            worksheet[cell].alignment = warp_text_alignment
+            worksheet[cell].value = beneficiary_type_columns[i%len(beneficiary_type_columns)]
+
+    def set_service_delivery_columns(start_column_index, row):
+        for i in range(6):
+            column_index = start_column_index + i*3
+            cell = "{}{}".format(columns[column_index], row)
+            worksheet[cell].fill = blue_fill
+            worksheet[cell].border = thin_border
+            worksheet[cell].font = bold_font
+            worksheet[cell].alignment = warp_text_alignment
+            worksheet[cell].value = secondary_headers[i]
+            next_cell = "{}{}".format(columns[column_index+2], row)
+            worksheet.merge_cells(f"{cell}:{next_cell}")
+
+    next_column_deviation = 0
+    for index, value in enumerate(headers):
+        column_index = index + next_column_deviation
+        cell = "{}{}".format(columns[column_index], table_header_position_row)
 
-    table_header = {}
-    for col, header in zip(columns, headers):
-        table_header[col] = header
-    for column, value in table_header.items():
-        cell = "{}{}".format(column, table_header_position_row)
         worksheet[cell].fill = grey_fill
         worksheet[cell].border = thin_border
         worksheet[cell].font = bold_font
         worksheet[cell].alignment = warp_text_alignment
         worksheet[cell].value = value
 
+        if report_type == 'beneficiary_wise':
+            if value in ('Total No. of Beneficiaries eligible for THR',
+                         'Total No. of beneficiaries received THR in given month'):
```

Comment:
I wish there could be more easy way to do it But all these different settings are for different Excel format asked for different report type:
you can check the spec in this ticket: https://app.asana.com/0/1175081284767805/1157414140082868/f, 
These +1, +2 are hardcoded for these report formats as these formattings are very much belonged to each report type specifically. I though tried to make the sharable formatting between those reports without duplicating code

Reply:
gotcha. might be nice to optionally add some context/comments so that someone looking at this later doesn't wonder the same thing... (although admittedly even after reading your comment I'm not clear exactly what they're doing... choosing a specific row in the sheet to put the data in?)